pkg/controlsvc/controlsvc.go

```diff
@@ -22,51 +24,117 @@ import (
 	"github.com/ghjm/cmdline"
 )
 
-// sockControl implements the ControlFuncOperations interface that is passed back to control functions.
-type sockControl struct {
-	conn net.Conn
+const (
+	normalCloseError         = "normal close"
+	writeControlServiceError = "Write error in control service"
+)
+
+type Copier interface {
+	Copy(dst io.Writer, src io.Reader) (written int64, err error)
+}
+
+type SocketConnIO struct{}
+
+func (s *SocketConnIO) Copy(dst io.Writer, src io.Reader) (written int64, err error) {
+	return io.Copy(dst, src)
+}
+
+type NetceptorForControlsvc interface {
+	ListenAndAdvertise(service string, tlscfg *tls.Config, tags map[string]string) (*netceptor.Listener, error)
+	NetceptorForControlCommand
+}
+
+type Utiler interface {
+	BridgeConns(c1 io.ReadWriteCloser, c1Name string, c2 io.ReadWriteCloser, c2Name string, logger *logger.ReceptorLogger)
+	UnixSocketListen(filename string, permissions fs.FileMode) (net.Listener, *utils.FLock, error)
+}
+
+type Util struct{}
+
+func (u *Util) BridgeConns(c1 io.ReadWriteCloser, c1Name string, c2 io.ReadWriteCloser, c2Name string, logger *logger.ReceptorLogger) {
+	utils.BridgeConns(c1, c1Name, c2, c2Name, logger)
+}
+
+func (u *Util) UnixSocketListen(filename string, permissions fs.FileMode) (net.Listener, *utils.FLock, error) {
+	return utils.UnixSocketListen(filename, permissions)
 }
 
-func (s *sockControl) RemoteAddr() net.Addr {
+type Neter interface {
+	Listen(network string, address string) (net.Listener, error)
+}
+
+type Net struct{}
+
+func (n *Net) Listen(network string, address string) (net.Listener, error) {
+	return net.Listen(network, address)
+}
+
+type Tlser interface {
+	NewListener(inner net.Listener, config *tls.Config) net.Listener
+}
+
+type Tls struct{}
+
+func (t *Tls) NewListener(inner net.Listener, config *tls.Config) net.Listener {
+	return tls.NewListener(inner, config)
+}
+
+// SockControl implements the ControlFuncOperations interface that is passed back to control functions.
+type SockControl struct {
+	conn  net.Conn
+	utils Utiler
+	io    Copier
+}
+
+func NewSockControl(conn net.Conn, utils Utiler, copier Copier) *SockControl {
+	return &SockControl{
+		conn:  conn,
+		utils: utils,
+		io:    copier,
+	}
+}
+
+func (s *SockControl) RemoteAddr() net.Addr {
 	return s.conn.RemoteAddr()
 }
 
-// BridgeConn bridges the socket to another socket.
-func (s *sockControl) BridgeConn(message string, bc io.ReadWriteCloser, bcName string, logger *logger.ReceptorLogger) error {
+// WriteMessage attempts to write a message to a connection
+func (s *SockControl) WriteMessage(message string) error {
 	if message != "" {
 		_, err := s.conn.Write([]byte(message))
 		if err != nil {
 			return err
 		}
 	}
-	utils.BridgeConns(s.conn, "control service", bc, bcName, logger)
+	return nil
+}
+
+// BridgeConn bridges the socket to another socket.
+func (s *SockControl) BridgeConn(message string, bc io.ReadWriteCloser, bcName string, logger *logger.ReceptorLogger) error {
```

Comment:
As part of the same comment for the fields in SockControl, and alternative would be that `BridgeConn` takes in a `utilier` argument